sql/byte_buffer.go

```diff
@@ -0,0 +1,57 @@
+package sql
+
+import (
+	"sync"
+)
+
+const defaultByteBuffCap = 4096
+
+var ByteBufPool = sync.Pool{
+	New: func() any {
+		return NewByteBuffer(defaultByteBuffCap)
+	},
+}
+
+type ByteBuffer struct {
+	i   int
+	buf []byte
+}
+
+func NewByteBuffer(initCap int) *ByteBuffer {
+	buf := make([]byte, initCap)
+	return &ByteBuffer{buf: buf}
+}
+
+// Grow records the latest used byte position. Callers
+// are responsible for accurately reporting which bytes
+// they expect to be protected.
+func (b *ByteBuffer) Grow(n int) {
+	if b.i+n > len(b.buf) {
```

Comment:
Actually is there an off-by-one error here? Test at the boundary would be goo

Reply:
yeah, not and error maybe but definitely better to preemptively double when we know the next Grow will exceed the buffer